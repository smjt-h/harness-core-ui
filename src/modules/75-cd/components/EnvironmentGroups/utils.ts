/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { defaultTo } from 'lodash-es'
import type {
  EnvironmentGroupFilterProperties,
  EnvironmentGroupResponseDTO,
  GetEnvironmentGroupListQueryParams
} from 'services/cd-ng'

export function cleanData(values: EnvironmentGroupResponseDTO): EnvironmentGroupResponseDTO {
  const {
    name: prevName,
    identifier: prevIdentifier,
    description: prevDescription,
    tags,
    orgIdentifier,
    projectIdentifier,
    envIdentifiers
  } = values

  const name = prevName?.toString().trim()
  const identifier = prevIdentifier?.toString().trim()
  const description = prevDescription?.toString().trim()
  return {
    name,
    identifier,
    description,
    tags: defaultTo(tags, {}),
    orgIdentifier,
    projectIdentifier,
    envIdentifiers
  } as any
}

export enum SortFields {
  LastUpdatedAt = 'lastModifiedAt',
  AZ09 = 'AZ09',
  ZA90 = 'ZA90',
  Name = 'name'
}

export enum Sort {
  DESC = 'DESC',
  ASC = 'ASC'
}

export enum EnvironmentGroupDetailsTab {
  CONFIGURATION = 'CONFIGURATION',
  ENVIRONMENTS = 'ENVIRONMENTS'
}

export interface EnvironmentGroupListQueryParams
  extends Partial<
    Omit<GetEnvironmentGroupListQueryParams, 'sort' | 'getDefaultFromOtherRepo' | 'envGroupIdentifiers'>
  > {
  filters?: EnvironmentGroupFilterProperties
}
