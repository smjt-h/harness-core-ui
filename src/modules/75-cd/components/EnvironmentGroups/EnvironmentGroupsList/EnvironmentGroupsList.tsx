/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { useState } from 'react'
import { useHistory, useParams } from 'react-router-dom'
import cx from 'classnames'
import { defaultTo } from 'lodash-es'

import { Button, Container, getErrorInfoFromErrorObject, Layout, Text, useToaster } from '@harness/uicore'
import { Color, FontVariation } from '@harness/design-system'
import { useStrings } from 'framework/strings'
import { EnvironmentGroupResponse, useDeleteEnvironmentGroup } from 'services/cd-ng'

import routes from '@common/RouteDefinitions'
import type { ModulePathParams, ProjectPathProps } from '@common/interfaces/RouteInterfaces'

import {
  EditOrDeleteCell,
  EnvironmentGroupName,
  EnvironmentTypes,
  LastUpdatedBy,
  NoOfEnvironments
} from './EnvironmentGroupsListColumns'
import { EnvironmentGroupDetailsTab } from '../utils'

import css from '../EnvironmentGroups.module.scss'

export interface EnvironmentGroupsListProps {
  environmentGroups?: EnvironmentGroupResponse[]
  refetch: () => void
}

export default function EnvironmentGroupsList({ environmentGroups, refetch }: EnvironmentGroupsListProps) {
  const { accountId, orgIdentifier, projectIdentifier, module } = useParams<ProjectPathProps & ModulePathParams>()
  const history = useHistory()
  const { getString } = useStrings()
  const { showSuccess, showError } = useToaster()
  const [expandedSets, setExpandedSets] = useState<Set<string>>(new Set())

  const onEdit = (environmentGroupIdentifier: string) => {
    history.push(
      routes.toEnvironmentGroupDetails({
        orgIdentifier,
        projectIdentifier,
        accountId,
        module,
        environmentGroupIdentifier,
        sectionId: EnvironmentGroupDetailsTab.CONFIGURATION
      })
    )
  }

  const { mutate: deleteEnvironmentGroup } = useDeleteEnvironmentGroup({
    queryParams: {
      accountIdentifier: accountId,
      projectIdentifier,
      orgIdentifier
    }
  })

  const onDelete = async (identifier: string) => {
    try {
      await deleteEnvironmentGroup(identifier, { headers: { 'content-type': 'application/json' } })
      showSuccess(getString('common.environmentGroup.deleted', { identifier }))
      refetch()
    } catch (e: any) {
      showError(getErrorInfoFromErrorObject(e))
    }
  }

  return (
    <>
      <EnvironmentGroupsListHeaders />
      {environmentGroups?.map(environmentGroup => {
        /* istanbul ignore next */ const {
          lastModifiedAt,
          envGroup: { name, identifier = '', tags, envResponse: environments } = {}
        } = environmentGroup
        return (
          <Container
            key={identifier}
            data-testid={identifier}
            width={'100%'}
            background={Color.WHITE}
            padding={{
              left: 'medium',
              top: 'medium',
              right: 'medium',
              bottom: expandedSets.has(identifier) ? 0 : 'medium'
            }}
            margin={{ bottom: 'small' }}
            className={cx(css.tableContainer, { [css.showPointer]: !expandedSets.has(identifier) })}
            onClick={() => {
              // istanbul ignore else
              if (!expandedSets.has(identifier)) {
                expandedSets.add(identifier)
                setExpandedSets(new Set(expandedSets))
              }
            }}
          >
            <Layout.Horizontal spacing="xsmall" flex={{ justifyContent: 'space-between' }}>
              <Layout.Vertical width={'60%'}>
                <EnvironmentGroupName name={name} identifier={identifier} tags={tags} />
              </Layout.Vertical>

              <Layout.Horizontal width={'40%'} flex={{ justifyContent: 'space-between' }}>
                <Container width={'120px'}>
                  <LastUpdatedBy lastModifiedAt={lastModifiedAt} />
                </Container>

                <NoOfEnvironments length={defaultTo(environments, []).length} />

                <EditOrDeleteCell identifier={identifier} onEdit={onEdit} onDelete={onDelete} />
              </Layout.Horizontal>
            </Layout.Horizontal>

            {expandedSets.has(identifier) && (
              <Layout.Vertical>
                {!environments?.length && (
                  <Text font={{ variation: FontVariation.BODY }} color={Color.GREY_500} padding={{ top: 'medium' }}>
                    {getString('cd.noEnvironment.title')}
                  </Text>
                )}
                {environments?.map(environment => {
                  /* istanbul ignore next */ const {
                    lastModifiedAt: envLastModifiedAt,
                    environment: { name: envName, identifier: envIdentifier, tags: envTags, type: envType } = {}
                  } = environment

                  return (
                    <Layout.Horizontal
                      spacing="xsmall"
                      padding={{ top: 'medium' }}
                      key={envIdentifier}
                      data-testid={envIdentifier}
                    >
                      <Layout.Vertical padding={{ left: 'xxlarge' }} style={{ flexGrow: 1 }}>
                        <EnvironmentGroupName name={envName} identifier={envIdentifier} tags={envTags} />
                      </Layout.Vertical>

                      <Layout.Horizontal flex={{ justifyContent: 'flex-start' }} width={'10%'}>
                        <EnvironmentTypes type={envType} />
                      </Layout.Horizontal>
                      <Layout.Horizontal width={'40%'} flex={{ justifyContent: 'flex-start' }}>
                        <LastUpdatedBy lastModifiedAt={envLastModifiedAt} />
                      </Layout.Horizontal>
                    </Layout.Horizontal>
                  )
                })}
                <Layout.Horizontal flex={{ align: 'center-center' }} margin={{ top: 'medium' }}>
                  <Button
                    rightIcon="chevron-up"
                    font={{ weight: 'semi-bold', variation: FontVariation.BODY1, size: 'small' }}
                    color={Color.GREY_100}
                    minimal
                    onClick={event => {
                      event.stopPropagation()
                      // istanbul ignore else
                      if (expandedSets.has(identifier)) {
                        expandedSets.delete(identifier)
                        setExpandedSets(new Set(expandedSets))
                      }
                    }}
                  >
                    {getString('common.showLess')}
                  </Button>
                </Layout.Horizontal>
              </Layout.Vertical>
            )}
          </Container>
        )
      })}
    </>
  )
}

function EnvironmentGroupsListHeaders() {
  const { getString } = useStrings()

  return (
    <Layout.Horizontal margin={{ top: 'large', bottom: 'medium' }}>
      <Text
        font={{ variation: FontVariation.TABLE_HEADERS }}
        style={{ flexGrow: 1 }}
        role="button"
        tabIndex={0}
        width={'50%'}
      >
        {getString('common.environmentGroup.label').toUpperCase()}
      </Text>
      <Text width={'10%'} font={{ variation: FontVariation.TABLE_HEADERS }}>
        {getString('typeLabel').toUpperCase()}
      </Text>
      <Text width={'40%'} font={{ variation: FontVariation.TABLE_HEADERS }} role="button" tabIndex={0}>
        {getString('lastUpdatedBy').toUpperCase()}
      </Text>
    </Layout.Horizontal>
  )
}
