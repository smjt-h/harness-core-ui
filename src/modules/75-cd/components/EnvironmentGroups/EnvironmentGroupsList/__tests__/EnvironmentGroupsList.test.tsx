/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { act, fireEvent, render, screen, waitFor } from '@testing-library/react'

import * as cdNgServices from 'services/cd-ng'

import routes from '@common/RouteDefinitions'
import { modulePathProps, projectPathProps } from '@common/utils/routeUtils'
import { findDialogContainer, findPopoverContainer, TestWrapper } from '@common/utils/testUtils'

import EnvironmentGroupsList from '../EnvironmentGroupsList'
import EnvironmentGroupsMockData from './__mocks__/environmentGroupsMockData.json'

describe('Environment Groups List View', () => {
  test('renders basic list and works like an accordion', async () => {
    const { container } = render(
      <TestWrapper
        path={routes.toEnvironmentGroups({ ...projectPathProps, ...modulePathProps })}
        pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy', module: 'cd' }}
      >
        <EnvironmentGroupsList
          environmentGroups={EnvironmentGroupsMockData as unknown as cdNgServices.EnvironmentGroupResponse[]}
          refetch={jest.fn()}
        />
      </TestWrapper>
    )

    expect(screen.queryByTestId('Env_1')).not.toBeInTheDocument()
    fireEvent.click(screen.getByTestId('Env_Group_4'))
    await waitFor(() => expect(screen.queryByTestId('Env_1')).toBeInTheDocument())

    fireEvent.click(screen.queryByText('common.showLess')!)
    expect(screen.queryByTestId('Env_1')).not.toBeInTheDocument()

    fireEvent.click(screen.getByTestId('Env_Group_2'))
    await waitFor(() => expect(screen.queryByText('cd.noEnvironment.title')).toBeInTheDocument())

    expect(container).toMatchSnapshot()
  })

  test('edit routes to details page', async () => {
    const { container } = render(
      <TestWrapper
        path={routes.toEnvironmentGroups({ ...projectPathProps, ...modulePathProps })}
        pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy', module: 'cd' }}
      >
        <EnvironmentGroupsList
          environmentGroups={EnvironmentGroupsMockData as unknown as cdNgServices.EnvironmentGroupResponse[]}
          refetch={jest.fn()}
        />
      </TestWrapper>
    )

    const moreIcon = container.querySelector('span[icon="more"]')
    fireEvent.click(moreIcon!)

    const popover = findPopoverContainer()
    const editMenuItem = popover?.querySelector('.bp3-icon-edit')
    fireEvent.click(editMenuItem!)

    expect(
      screen.getByTestId('location').innerHTML.endsWith(
        routes.toEnvironmentGroupDetails({
          orgIdentifier: 'dummy',
          projectIdentifier: 'dummy',
          accountId: 'dummy',
          module: 'cd',
          environmentGroupIdentifier: 'Env_Group_4',
          sectionId: 'CONFIGURATION'
        })
      )
    ).toEqual(true)
  })

  test('delete env group asks for confirmation before deleting', async () => {
    const mutate = jest.fn()
    jest.spyOn(cdNgServices, 'useDeleteEnvironmentGroup').mockImplementation(
      () =>
        ({
          loading: false,
          mutate,
          cancel: jest.fn(),
          error: null
        } as any)
    )
    const { container } = render(
      <TestWrapper
        path={routes.toEnvironmentGroups({ ...projectPathProps, ...modulePathProps })}
        pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy', module: 'cd' }}
      >
        <EnvironmentGroupsList
          environmentGroups={EnvironmentGroupsMockData as unknown as cdNgServices.EnvironmentGroupResponse[]}
          refetch={jest.fn()}
        />
      </TestWrapper>
    )

    const moreIcon = container.querySelector('span[icon="more"]')
    fireEvent.click(moreIcon!)

    const deleteMenuItem = findPopoverContainer()?.querySelector('.bp3-icon-trash')
    fireEvent.click(deleteMenuItem!)

    expect(findDialogContainer()).not.toBeNull()

    const cancelButton = findDialogContainer()?.querySelectorAll('.bp3-button-text')[1]
    fireEvent.click(cancelButton!)

    expect(findDialogContainer()).toBeNull()
    expect(mutate).not.toHaveBeenCalled()

    fireEvent.click(moreIcon!)
    fireEvent.click(deleteMenuItem!)

    act(() => {
      const deleteButton = findDialogContainer()?.querySelectorAll('.bp3-button-text')[0]
      fireEvent.click(deleteButton!)
    })

    await waitFor(() =>
      expect(mutate).toHaveBeenCalledWith('Env_Group_4', { headers: { 'content-type': 'application/json' } })
    )
  })

  test('delete env group fails and displays toast', async () => {
    const mutate = jest.fn().mockRejectedValue({ message: 'Failed to delete' })
    jest.spyOn(cdNgServices, 'useDeleteEnvironmentGroup').mockImplementation(
      () =>
        ({
          loading: false,
          mutate,
          cancel: jest.fn()
        } as any)
    )
    const { container } = render(
      <TestWrapper
        path={routes.toEnvironmentGroups({ ...projectPathProps, ...modulePathProps })}
        pathParams={{ accountId: 'dummy', orgIdentifier: 'dummy', projectIdentifier: 'dummy', module: 'cd' }}
      >
        <EnvironmentGroupsList
          environmentGroups={EnvironmentGroupsMockData as unknown as cdNgServices.EnvironmentGroupResponse[]}
          refetch={jest.fn()}
        />
      </TestWrapper>
    )

    const moreIcon = container.querySelector('span[icon="more"]')
    fireEvent.click(moreIcon!)

    const deleteMenuItem = findPopoverContainer()?.querySelector('.bp3-icon-trash')
    fireEvent.click(deleteMenuItem!)

    expect(findDialogContainer()).not.toBeNull()

    act(() => {
      const deleteButton = findDialogContainer()?.querySelectorAll('.bp3-button-text')[0]
      fireEvent.click(deleteButton!)
    })

    await waitFor(() => expect(screen.queryByText('Failed to delete')).toBeInTheDocument())
  })
})
