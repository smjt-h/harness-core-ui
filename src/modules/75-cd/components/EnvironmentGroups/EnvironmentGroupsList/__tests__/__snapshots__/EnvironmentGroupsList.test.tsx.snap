// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Environment Groups List View renders basic list and works like an accordion 1`] = `
<div>
  <div
    class="Layout--horizontal StyledProps--main StyledProps--margin-top-large StyledProps--margin-bottom-medium"
  >
    <p
      class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
      role="button"
      style="flex-grow: 1; width: 50%;"
      tabindex="0"
    >
      COMMON.ENVIRONMENTGROUP.LABEL
    </p>
    <p
      class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
      style="width: 10%;"
    >
      TYPELABEL
    </p>
    <p
      class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
      role="button"
      style="width: 40%;"
      tabindex="0"
    >
      LASTUPDATEDBY
    </p>
  </div>
  <div
    class="StyledProps--font StyledProps--main tableContainer showPointer StyledProps--background-white StyledProps--padding-left-medium StyledProps--padding-top-medium StyledProps--padding-right-medium StyledProps--padding-bottom-medium StyledProps--margin-bottom-small"
    data-testid="Env_Group_4"
    style="width: 100%;"
  >
    <div
      class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-space-between"
    >
      <div
        class="Layout--vertical StyledProps--main"
        style="width: 60%;"
      >
        <div
          class="Layout--horizontal Layout--layout-spacing-small StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-flex-start StyledProps--margin-bottom-small"
        >
          <p
            class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--color StyledProps--color-black"
            style="--text-line-clamp: 1;"
          >
            Env Group 4
          </p>
          <span
            class="bp3-popover-wrapper"
          >
            <span
              class="bp3-popover-target"
            >
              <div
                class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main tagsPopover StyledProps--flex StyledProps--flex-align-center-center"
                tabindex="0"
              >
                <span
                  class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-grey600"
                  data-icon="main-tags"
                >
                  <svg
                    height="14"
                    viewBox="0 0 14 14"
                    width="14"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M7 0h-.2c-.2 0-.4.1-.6.3l-6 6c-.3.3-.3.8 0 1.1L2 9.1l4.6 4.6c.3.3.8.3 1.1 0l6-6c.2-.1.3-.3.3-.5v-.7-.3-1.3-2c0-.5-.4-1-1-1h-.9V1c0-.5-.4-1-1-1H7zm0 .6h4.2c.2 0 .4.2.4.4v.9h-1.1c-.3-.3-.5-.4-.8-.4s-.4.1-.6.3h-.5c-.2.1-.4.1-.6.3l-6 6s-.1.1-.1.2L.7 6.9c-.1-.1-.1-.2 0-.3l6-6s.1-.1.2-.1c0 .1.1.1.1.1zm1.9 1.8H13c.2 0 .4.2.4.4v4.3s0 .2-.1.2l-6 6c-.1.1-.2.1-.3 0L2.5 8.8c-.1-.1-.1-.2 0-.3l6-6s.1-.1.2-.1zm2.7.9c-.2 0-.5.1-.6.3-.4.4-.4.9 0 1.3s.9.4 1.3 0 .4-.9 0-1.3c-.3-.2-.5-.3-.7-.3zm0 .7c.1 0 .2 0 .2.1.1.1.1.3 0 .4s-.3.1-.4 0-.1-.3 0-.4c0-.1.1-.1.2-.1z"
                      fill="#4d5969"
                    />
                  </svg>
                </span>
                <p
                  class="StyledProps--font StyledProps--main"
                >
                  1
                </p>
              </div>
            </span>
          </span>
        </div>
        <p
          class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--color StyledProps--color-grey500 StyledProps--font-size-small"
          style="--text-line-clamp: 1;"
        >
          common.ID
          : 
          Env_Group_4
        </p>
      </div>
      <div
        class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-space-between"
        style="width: 40%;"
      >
        <div
          class="StyledProps--font StyledProps--main"
          style="width: 120px;"
        >
          <div
            class="Layout--vertical Layout--layout-spacing-small StyledProps--main"
          >
            dummy date
          </div>
        </div>
        <span
          class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-black StyledProps--background-grey100 StyledProps--padding-left-large StyledProps--padding-top-small StyledProps--padding-right-large StyledProps--padding-bottom-small StyledProps--inline"
        >
          5
           environment(s) included
        </span>
        <div
          class="Layout--horizontal StyledProps--main"
          style="justify-content: flex-end;"
        >
          <span
            class="bp3-popover-wrapper bp3-dark"
          >
            <span
              class="bp3-popover-target"
            >
              <button
                class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly Button--without-shadow Button--withLeftIcon"
                style="transform: rotate(90deg);"
                type="button"
              >
                <span
                  class="bp3-icon bp3-icon-more StyledProps--main"
                  icon="more"
                >
                  <svg
                    data-icon="more"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <desc>
                      more
                    </desc>
                    <path
                      d="M2 6.03a2 2 0 100 4 2 2 0 100-4zM14 6.03a2 2 0 100 4 2 2 0 100-4zM8 6.03a2 2 0 100 4 2 2 0 100-4z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </button>
            </span>
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    class="StyledProps--font StyledProps--main tableContainer StyledProps--background-white StyledProps--padding-left-medium StyledProps--padding-top-medium StyledProps--padding-right-medium StyledProps--margin-bottom-small"
    data-testid="Env_Group_2"
    style="width: 100%;"
  >
    <div
      class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-space-between"
    >
      <div
        class="Layout--vertical StyledProps--main"
        style="width: 60%;"
      >
        <div
          class="Layout--horizontal Layout--layout-spacing-small StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-flex-start StyledProps--margin-bottom-small"
        >
          <p
            class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--color StyledProps--color-black"
            style="--text-line-clamp: 1;"
          >
            Env Group 2
          </p>
          <span
            class="bp3-popover-wrapper"
          >
            <span
              class="bp3-popover-target"
            >
              <div
                class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main tagsPopover StyledProps--flex StyledProps--flex-align-center-center"
                tabindex="0"
              >
                <span
                  class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-grey600"
                  data-icon="main-tags"
                >
                  <svg
                    height="14"
                    viewBox="0 0 14 14"
                    width="14"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M7 0h-.2c-.2 0-.4.1-.6.3l-6 6c-.3.3-.3.8 0 1.1L2 9.1l4.6 4.6c.3.3.8.3 1.1 0l6-6c.2-.1.3-.3.3-.5v-.7-.3-1.3-2c0-.5-.4-1-1-1h-.9V1c0-.5-.4-1-1-1H7zm0 .6h4.2c.2 0 .4.2.4.4v.9h-1.1c-.3-.3-.5-.4-.8-.4s-.4.1-.6.3h-.5c-.2.1-.4.1-.6.3l-6 6s-.1.1-.1.2L.7 6.9c-.1-.1-.1-.2 0-.3l6-6s.1-.1.2-.1c0 .1.1.1.1.1zm1.9 1.8H13c.2 0 .4.2.4.4v4.3s0 .2-.1.2l-6 6c-.1.1-.2.1-.3 0L2.5 8.8c-.1-.1-.1-.2 0-.3l6-6s.1-.1.2-.1zm2.7.9c-.2 0-.5.1-.6.3-.4.4-.4.9 0 1.3s.9.4 1.3 0 .4-.9 0-1.3c-.3-.2-.5-.3-.7-.3zm0 .7c.1 0 .2 0 .2.1.1.1.1.3 0 .4s-.3.1-.4 0-.1-.3 0-.4c0-.1.1-.1.2-.1z"
                      fill="#4d5969"
                    />
                  </svg>
                </span>
                <p
                  class="StyledProps--font StyledProps--main"
                >
                  1
                </p>
              </div>
            </span>
          </span>
        </div>
        <p
          class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--color StyledProps--color-grey500 StyledProps--font-size-small"
          style="--text-line-clamp: 1;"
        >
          common.ID
          : 
          Env_Group_2
        </p>
      </div>
      <div
        class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-space-between"
        style="width: 40%;"
      >
        <div
          class="StyledProps--font StyledProps--main"
          style="width: 120px;"
        >
          <div
            class="Layout--vertical Layout--layout-spacing-small StyledProps--main"
          >
            dummy date
          </div>
        </div>
        <span
          class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-black StyledProps--background-grey100 StyledProps--padding-left-large StyledProps--padding-top-small StyledProps--padding-right-large StyledProps--padding-bottom-small StyledProps--inline"
        >
          0
           environment(s) included
        </span>
        <div
          class="Layout--horizontal StyledProps--main"
          style="justify-content: flex-end;"
        >
          <span
            class="bp3-popover-wrapper bp3-dark"
          >
            <span
              class="bp3-popover-target"
            >
              <button
                class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly Button--without-shadow Button--withLeftIcon"
                style="transform: rotate(90deg);"
                type="button"
              >
                <span
                  class="bp3-icon bp3-icon-more StyledProps--main"
                  icon="more"
                >
                  <svg
                    data-icon="more"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <desc>
                      more
                    </desc>
                    <path
                      d="M2 6.03a2 2 0 100 4 2 2 0 100-4zM14 6.03a2 2 0 100 4 2 2 0 100-4zM8 6.03a2 2 0 100 4 2 2 0 100-4z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </button>
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="Layout--vertical StyledProps--main"
    >
      <p
        class="StyledProps--font StyledProps--main StyledProps--font-variation-body StyledProps--color StyledProps--color-grey500 StyledProps--padding-top-medium"
      >
        cd.noEnvironment.title
      </p>
      <div
        class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-align-center-center StyledProps--margin-top-medium"
      >
        <button
          class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main StyledProps--font StyledProps--font-weight-semi-bold StyledProps--font-variation-body1 StyledProps--font-size-small StyledProps--color StyledProps--color-grey100 Button--with-current-color Button--iconOnly Button--without-shadow Button--withRightIcon"
          type="button"
        >
          <span
            class="bp3-button-text"
          >
            common.showLess
          </span>
          <span
            class="bp3-icon bp3-icon-chevron-up StyledProps--main"
            icon="chevron-up"
          >
            <svg
              data-icon="chevron-up"
              height="16"
              viewBox="0 0 16 16"
              width="16"
            >
              <desc>
                chevron-up
              </desc>
              <path
                d="M12.71 9.29l-4-4C8.53 5.11 8.28 5 8 5s-.53.11-.71.29l-4 4a1.003 1.003 0 001.42 1.42L8 7.41l3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
  <div
    class="StyledProps--font StyledProps--main tableContainer showPointer StyledProps--background-white StyledProps--padding-left-medium StyledProps--padding-top-medium StyledProps--padding-right-medium StyledProps--padding-bottom-medium StyledProps--margin-bottom-small"
    data-testid="Env_Group_3"
    style="width: 100%;"
  >
    <div
      class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-space-between"
    >
      <div
        class="Layout--vertical StyledProps--main"
        style="width: 60%;"
      >
        <div
          class="Layout--horizontal Layout--layout-spacing-small StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-flex-start StyledProps--margin-bottom-small"
        >
          <p
            class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--color StyledProps--color-black"
            style="--text-line-clamp: 1;"
          >
            Env Group 3
          </p>
        </div>
        <p
          class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--color StyledProps--color-grey500 StyledProps--font-size-small"
          style="--text-line-clamp: 1;"
        >
          common.ID
          : 
          Env_Group_3
        </p>
      </div>
      <div
        class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-space-between"
        style="width: 40%;"
      >
        <div
          class="StyledProps--font StyledProps--main"
          style="width: 120px;"
        >
          <div
            class="Layout--vertical Layout--layout-spacing-small StyledProps--main"
          >
            dummy date
          </div>
        </div>
        <span
          class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-black StyledProps--background-grey100 StyledProps--padding-left-large StyledProps--padding-top-small StyledProps--padding-right-large StyledProps--padding-bottom-small StyledProps--inline"
        >
          2
           environment(s) included
        </span>
        <div
          class="Layout--horizontal StyledProps--main"
          style="justify-content: flex-end;"
        >
          <span
            class="bp3-popover-wrapper bp3-dark"
          >
            <span
              class="bp3-popover-target"
            >
              <button
                class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly Button--without-shadow Button--withLeftIcon"
                style="transform: rotate(90deg);"
                type="button"
              >
                <span
                  class="bp3-icon bp3-icon-more StyledProps--main"
                  icon="more"
                >
                  <svg
                    data-icon="more"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <desc>
                      more
                    </desc>
                    <path
                      d="M2 6.03a2 2 0 100 4 2 2 0 100-4zM14 6.03a2 2 0 100 4 2 2 0 100-4zM8 6.03a2 2 0 100 4 2 2 0 100-4z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </button>
            </span>
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    class="StyledProps--font StyledProps--main tableContainer showPointer StyledProps--background-white StyledProps--padding-left-medium StyledProps--padding-top-medium StyledProps--padding-right-medium StyledProps--padding-bottom-medium StyledProps--margin-bottom-small"
    data-testid="Env_Group_1"
    style="width: 100%;"
  >
    <div
      class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-space-between"
    >
      <div
        class="Layout--vertical StyledProps--main"
        style="width: 60%;"
      >
        <div
          class="Layout--horizontal Layout--layout-spacing-small StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-flex-start StyledProps--margin-bottom-small"
        >
          <p
            class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--color StyledProps--color-black"
            style="--text-line-clamp: 1;"
          >
            Env Group 1
          </p>
        </div>
        <p
          class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--color StyledProps--color-grey500 StyledProps--font-size-small"
          style="--text-line-clamp: 1;"
        >
          common.ID
          : 
          Env_Group_1
        </p>
      </div>
      <div
        class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-justifyContent-space-between"
        style="width: 40%;"
      >
        <div
          class="StyledProps--font StyledProps--main"
          style="width: 120px;"
        >
          <div
            class="Layout--vertical Layout--layout-spacing-small StyledProps--main"
          >
            dummy date
          </div>
        </div>
        <span
          class="StyledProps--font StyledProps--main StyledProps--color StyledProps--color-black StyledProps--background-grey100 StyledProps--padding-left-large StyledProps--padding-top-small StyledProps--padding-right-large StyledProps--padding-bottom-small StyledProps--inline"
        >
          2
           environment(s) included
        </span>
        <div
          class="Layout--horizontal StyledProps--main"
          style="justify-content: flex-end;"
        >
          <span
            class="bp3-popover-wrapper bp3-dark"
          >
            <span
              class="bp3-popover-target"
            >
              <button
                class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly Button--without-shadow Button--withLeftIcon"
                style="transform: rotate(90deg);"
                type="button"
              >
                <span
                  class="bp3-icon bp3-icon-more StyledProps--main"
                  icon="more"
                >
                  <svg
                    data-icon="more"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <desc>
                      more
                    </desc>
                    <path
                      d="M2 6.03a2 2 0 100 4 2 2 0 100-4zM14 6.03a2 2 0 100 4 2 2 0 100-4zM8 6.03a2 2 0 100 4 2 2 0 100-4z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </button>
            </span>
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
